### PR TITLE
fix(app): harden timing-safe verification for feishu and tool leases

### LINKED ISSUE #428: [RFC] Timing-Safe Comparison & HMAC Webhook Verification
# [RFC] Timing-Safe Comparison & HMAC Webhook Verification

**Labels**: `enhancement`, `priority:P0`, `mvp`
**Effort**: S
**Date**: 2026-03-21

---

## Problem Statement

Loongclaw has three verification paths that use non-constant-time string
comparison (`!=`), exposing each to timing side-channel attacks:

1. **`verify_feishu_token`** (`crates/app/src/channel/feishu/payload/inbound.rs:219`):
   Compares a plaintext verification token from the incoming payload against the
   configured token using `if incoming != expected_token`. An attacker can submit
   requests with progressively refined token guesses, measuring response latency to
   determine how many leading bytes match. With sufficient attempts, the full token
   can be recovered byte-by-byte.

2. **`verify_feishu_signature`** (`crates/app/src/channel/feishu/webhook.rs:783`):
   Computes SHA-256 over `timestamp || nonce || encrypt_key || body`, formats the
   digest as hex, then compares it against the `X-Lark-Signature` header using
   `if expected != signature`. Although the SHA-256 computation itself is not
   vulnerable, the hex-string comparison leaks information about how many leading
   characters of the expected digest match the attacker-supplied value. This is the
   more critical instance: timing attacks on hash/HMAC comparisons are
   well-documented (OWASP, CWE-208).

3. **`validate_tool_lease`** (`crates/app/src/tools/mod.rs:1357`):
   Computes SHA-256 over `tool_lease_secret || ":" || encoded_claims`, formats
   the digest as hex, then compares it against the signature portion of the lease
   using `if expected_signature != signature`. Same vulnerability class as (2).
   An attacker who recovers the lease signature can forge tool-call leases,
   bypassing tool authorization controls.

Additionally, Loongclaw has no HMAC-based webhook signature verification. Feishu
uses a raw `SHA-256(timestamp + nonce + encrypt_key + body)` scheme, but as more
channels are added (Slack uses HMAC-SHA256, Discord uses Ed25519, GitHub uses
HMAC-SHA256), each will need a standardized webhook verification utility.
Implementing this ad-hoc per channel will lead to repeated timing-unsafe patterns.

### Impact

- **CWE-208** (Observable Timing Discrepancy): All three verification paths are
  vulnerable to remote timing attacks.
- **OWASP**: "Use constant-time comparison functions for all secret comparisons"
  (OWASP Cryptographic Failures, A02:2021).
- **Blast radius**: Webhook endpoints are externally reachable. An attacker who
  recovers the verification token can forge arbitrary webhook payloads, injecting
  messages into any connected Feishu conversation. An attacker who recovers tool
  lease signatures can forge tool-call authorization, bypassing policy controls.

---

### 1. Add `subtle` crate as a direct dependency

The `subtle` crate (RustCrypto project) provides `ConstantTimeEq`, the standard
Rust primitive for constant-time byte comparison. It is already present in
`Cargo.lock` as a transitive dependency of `curve25519-dalek`, `digest`,
`ed25519-dalek`, and `rustls`. Adding it as a direct dependency introduces zero
new supply-chain surface.

### 2. Create `crypto::timing_safe_eq()` utility

A small, well-tested utility function that wraps `subtle::ConstantTimeEq` for
ergonomic use with string/byte-slice comparisons. This becomes the single
approved way to compare secrets anywhere in the codebase.

### 3. Add HMAC-SHA256 verification function

A generic `crypto::verify_hmac_sha256()` function that takes a secret key, a
message, and an expected signature, computes the HMAC, and compares using
constant-time equality. This covers Slack, GitHub, LINE, and any future
HMAC-based webhook scheme.

### 4. Refactor existing verification sites

Replace all three `!=` comparison sites with calls to the new utility functions.
No behavioral change -- only the comparison operator changes.

---

### Module location

```
crates/app/src/crypto/
    mod.rs          -- pub mod timing;  pub mod hmac;
    timing.rs       -- timing_safe_eq()
    hmac.rs         -- verify_hmac_sha256(), compute_hmac_sha256()
```

`crates/app/src/crypto/` is chosen because:
- All three current consumers (`webhook.rs`, `inbound.rs`, `tools/mod.rs`) are
  in `crates/app/`.
- The existing `crates/app/src/channel/feishu/payload/crypto.rs` (AES-256-CBC
  decryption) is domain-specific to Feishu and should remain where it is.
- A shared `crypto/` module at the `app` crate level is reachable by all channel
  adapters and tool modules added in the future.
- If a second crate needs these primitives later, they can be extracted to a
  `loongclaw-crypto` crate without breaking the API.

### `timing_safe_eq(a, b) -> bool`

```rust
use subtle::ConstantTimeEq;

/// Compares two byte sequences in constant time.
///
/// Returns `true` if `a` and `b` are equal. When lengths differ, returns
/// `false` immediately. In all concrete use sites (fixed-length tokens,
/// fixed-length hex digests), both inputs are the same length, so the
/// equal-length path is the security-critical one.
pub fn timing_safe_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    a.ct_eq(b).into()
}
```

Design notes:
- Length check is explicit. When lengths differ, we return `false` directly.
  This early return *does* leak the fact that lengths differ, but this is
  acceptable for our use cases: Feishu verification tokens are fixed-length,
  SHA-256 hex digests are always 64 characters, and tool lease signatures are
  always 64 characters. An attacker who supplies the wrong length learns nothing
  they did not already know. This is the same approach used by Go's
  `subtle.ConstantTimeCompare` and Ring's `constant_time::verify_slices_are_equal`.
- The `subtle::ConstantTimeEq` trait's `ct_eq` method on `[u8]` slices requires
  equal-length inputs (it panics on length mismatch in debug builds), so the
  length guard is mandatory, not optional.
- If a future use case requires hiding length differences (unlikely), the caller
  should normalize inputs to a fixed length before calling this function, or use
  an HMAC-then-compare pattern which inherently produces fixed-length outputs.

### `verify_hmac_sha256(key, message, expected_signature) -> bool`

```rust
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

/// Verifies an HMAC-SHA256 signature in constant time.
///
/// `expected_hex` is the hex-encoded expected signature (e.g., from an
/// HTTP header). Returns `true` if the computed HMAC matches.
pub fn verify_hmac_sha256(key: &[u8], message: &[u8], expected_hex: &str) -> bool {
    let Ok(expected_bytes) = hex::decode(expected_hex) else {
        return false;
    };
    let mut mac = HmacSha256::new_from_slice(key)
        .expect("HMAC-SHA256 accepts any key length");
    mac.update(message);
    // `mac.verify_slice` uses constant-time comparison internally
    // (via `digest` -> `subtle::ConstantTimeEq`).
    mac.verify_slice(&expected_bytes).is_ok()
}
```

Design notes:
- The `hmac` crate's `verify_slice` delegates to `digest::MacError` which uses
  `subtle::ConstantTimeEq` internally, so we get constant-time comparison for
  free. Verified: `hmac 0.12.1` depends on `digest 0.10.7`, which depends on
  `subtle 2.6.1`.
- Hex decoding failure returns `false` (no panic on malformed input).
- This function is generic enough for Slack (`v0=sha256hex`), GitHub
  (`sha256=hex`), and LINE (`base64` variant can get a sibling function).

### HMAC verification flow (per channel)

```
                  Incoming HTTP Request
                         |
          +--------------+--------------+
          |              |              |
    Extract header   Extract body   Extract timestamp
    (e.g. X-Slack-   (raw bytes)    (if replay
     Signature)                      protection needed)
          |              |              |
          +--------------+--------------+
                         |
              Canonicalize message
              (channel-specific:
               Slack: "v0:{ts}:{body}"
               GitHub: body as-is
               Feishu: ts+nonce+key+body)
                         |
              Compute HMAC-SHA256
              (or plain SHA-256 for Feishu)
                         |
              Constant-time compare
              against header value
                         |
                   +-----+-----+
                   |           |
                200 OK    401 Unauthorized
```

### Refactoring existing code

**Site 1: `verify_feishu_token` (inbound.rs:219)**:

```rust
// BEFORE:
if incoming != expected_token {
    return Err("unauthorized: feishu verification token mismatch".to_owned());
}

// AFTER:
if !crate::crypto::timing::timing_safe_eq(
    incoming.as_bytes(),
    expected_token.as_bytes(),
) {
    return Err("unauthorized: feishu verification token mismatch".to_owned());
}
```

**Site 2: `verify_feishu_signature` (webhook.rs:783)**:

```rust
// BEFORE:
let expected = format!("{:x}", hasher.finalize());
if expected != signature {
    return Err(( ... ));
}

// AFTER:
let expected = format!("{:x}", hasher.finalize());
if !crate::crypto::timing::timing_safe_eq(
    expected.as_bytes(),
    signature.as_bytes(),
) {
    return Err(( ... ));
}
```

**Site 3: `validate_tool_lease` (tools/mod.rs:1357)**:

```rust
// BEFORE:
let expected_signature = sign_tool_lease(encoded_claims);
if expected_signature != signature {
    return Err("invalid_tool_lease: signature mismatch".to_owned());
}

// AFTER:
let expected_signature = sign_tool_lease(encoded_claims);
if !crate::crypto::timing::timing_safe_eq(
    expected_signature.as_bytes(),
    signature.as_bytes(),
) {
    return Err("invalid_tool_lease: signature mismatch".to_owned());
}
```

Note: Both Feishu's signature scheme and the tool lease scheme use raw SHA-256,
not HMAC. The refactoring uses `timing_safe_eq` directly rather than
`verify_hmac_sha256`. The HMAC function is provided for future channels (Slack,
GitHub).

### New dependencies

| Crate    | Version | Purpose                        | Supply chain note                                                      |
|----------|---------|--------------------------------|------------------------------------------------------------------------|
| `subtle` | 2.x     | `ConstantTimeEq` trait         | Already in `Cargo.lock` via `curve25519-dalek`, `digest`, `rustls`     |
| `hmac`   | 0.12.x  | HMAC computation               | Already in `Cargo.lock` via `aws-sigv4`                                |
| `hex`    | 0.4.x   | Hex decode for HMAC sig verify | Already in `Cargo.lock` via `aws-sigv4`; not a direct dep of `crates/app` yet |

`sha2` is already a direct dependency of `crates/app` (workspace dependency).

Note: `subtle` and `hmac` are not direct dependencies of `crates/app` today --
they are only present as transitive dependencies of other crates. Adding them as
direct dependencies in `crates/app/Cargo.toml` pins the versions explicitly and
makes the dependency intentional rather than accidental. If the transitive
dependency chain changes (e.g., `aws-sigv4` drops `hmac`), the direct dependency
ensures continued availability.

---

## Acceptance Criteria

- [ ] `crypto::timing::timing_safe_eq()` exists and is covered by unit tests
      confirming: equal inputs return `true`, differing inputs return `false`,
      differing lengths return `false`.
- [ ] `crypto::hmac::verify_hmac_sha256()` exists and is covered by unit tests
      using known test vectors (RFC 4231 test vectors for HMAC-SHA256).
- [ ] `verify_feishu_token` (inbound.rs:219) uses `timing_safe_eq` instead of `!=`.
- [ ] `verify_feishu_signature` (webhook.rs:783) uses `timing_safe_eq` instead of `!=`.
- [ ] `validate_tool_lease` (tools/mod.rs:1357) uses `timing_safe_eq` instead of `!=`.
- [ ] All existing Feishu webhook tests and tool lease tests pass without
      modification (the change is transparent to callers).
- [ ] `subtle`, `hmac`, and `hex` are declared as direct dependencies in
      `crates/app/Cargo.toml`.
- [ ] No other code path in `crates/` compares secrets using `==` or `!=`
      (verified via grep for patterns like `token != `, `secret == `,
      `signature != `, etc.).
- [ ] `cargo clippy` and `cargo test` pass.

---

## Scope

This is a small, focused security hardening. Estimated effort: **S** (a few
hours of implementation plus review).

### In scope
- `timing_safe_eq()` utility
- `verify_hmac_sha256()` utility
- Refactoring the three existing verification functions
- Unit tests for the new utilities

### Out of scope
- Replacing Feishu's SHA-256 scheme with HMAC (that is Feishu's protocol, not ours)
- Replacing the tool lease SHA-256 scheme with HMAC (separate design decision)
- Implementing Slack/Discord/GitHub webhook verification (those are separate issues
  that will consume the utilities built here)
- `zeroize` for secret memory (separate P0 issue)
- `SecretRef` abstraction (separate P0 issue, much larger scope)

---

## References

| Source | Relevance |
|--------|-----------|
| Gap analysis: `_gap-analysis/17-secrets-auth/gap-matrix.md` (rows P0-3, P0-4) | Identifies the timing-safe comparison gap and HMAC verification gap |
| Challenger review: `_gap-analysis/reviews/17-secrets-auth-review.md` (sections P0-3, P0-4) | Confirms Feishu `!=` sites with exact line numbers |
| OWASP A02:2021 Cryptographic Failures | "Use constant-time comparison functions for all secret comparisons" |
| CWE-208: Observable Timing Discrepancy | Classification of the vulnerability |
| RFC 4231: HMAC-SHA-256 Test Vectors | Canonical test vectors for validating HMAC implementation |
| `subtle` crate (RustCrypto) | Rust ecosystem standard for constant-time operations |
| `hmac` crate (RustCrypto) | Rust ecosystem standard for HMAC computation |
| Existing code: `crates/app/src/channel/feishu/payload/inbound.rs:194-223` | `verify_feishu_token` -- plaintext token comparison with `!=` at line 219 |
| Existing code: `crates/app/src/channel/feishu/webhook.rs:755-790` | `verify_feishu_signature` -- SHA-256 hex digest comparison with `!=` at line 783 |
| Existing code: `crates/app/src/tools/mod.rs:1348-1359` | `validate_tool_lease` -- SHA-256 hex digest comparison with `!=` at line 1357 |

---

## Changelog (review corrections applied 2026-03-21)

1. **Added missing vulnerable site**: `validate_tool_lease` in
   `crates/app/src/tools/mod.rs:1357` uses the same `!=` comparison on a SHA-256
   hex digest. The original RFC listed only two sites; there are three.

2. **Corrected transitive dependency attribution for `subtle`**: The original RFC
   claimed `subtle` was a transitive dep of `aws-lc-rs`. Verified in
   `Cargo.lock`: `aws-lc-rs 1.16.1` depends only on `aws-lc-sys` and `zeroize`,
   not `subtle`. The actual transitive parents are `curve25519-dalek`, `digest`,
   `ed25519-dalek`, and `rustls`.

3. **Corrected transitive dependency attribution for `hmac`**: The original RFC
   did not specify which crate pulls in `hmac`. Verified: only `aws-sigv4`
   depends on `hmac` in `Cargo.lock`. This is a narrow dependency chain; if
   `aws-sigv4` is ever removed, `hmac` would disappear from the lock file.
   Adding it as a direct dep is therefore important, not just cosmetic.

4. **Fixed `timing_safe_eq` length-mismatch handling**: The original RFC's
   implementation ran `a.ct_eq(a)` as a "dummy burn" when lengths differed and
   claimed this "prevents length-based timing leaks." This was incorrect:
   - The burn time is proportional to `len(a)` (the secret), so it actually
     *leaks* information about the secret's length to an attacker who controls
     `len(b)`.
   - `subtle::ConstantTimeEq::ct_eq` on `[u8]` panics when slice lengths
     differ (in debug builds), so the length guard is mandatory, not a
     defense-in-depth choice.
   - The corrected implementation returns `false` directly on length mismatch,
     consistent with Go's `subtle.ConstantTimeCompare` and Ring's
     `constant_time::verify_slices_are_equal`. This is safe because all three
     use sites compare fixed-length values (tokens, 64-char hex digests).

5. **Clarified `hex` crate status**: The original RFC said "evaluate" for `hex`.
   Verified: `hex 0.4.3` is already in `Cargo.lock` (via `aws-sigv4`) but is
   not a direct dep of `crates/app`. It should be added as a direct dep alongside
   `subtle` and `hmac`. `format!("{:x}", ...)` cannot replace `hex::decode()`
   since the HMAC verification function needs to *decode* incoming hex strings.

6. **Updated acceptance criteria**: Added criterion for `validate_tool_lease`.
   Added `hex` to the direct dependency requirement. Updated "two" to "three"
   throughout.

7. **Corrected inbound.rs line range in References**: The function spans lines
   194-223 (not 194-222; line 223 is the closing brace of the function).

### PR DESCRIPTION
**Pull request by chumyin**
## Summary

- Problem: Several webhook and lease verification paths in `loongclaw-app` compared secrets and signatures with plain equality, leaving issue `#428` partially unaddressed.
- Why it matters: Authentication checks should avoid timing-sensitive comparisons and benefit from a reusable HMAC verification helper instead of ad hoc string equality. The PR also needs to stay green against current `dev`, including the browser-without-webfetch build matrix and security audit gate.
- What changed: Added shared `crypto` helpers for constant-time equality and HMAC-SHA256 verification, then migrated Feishu token verification, Feishu signature verification, and tool lease signature verification to the shared timing-safe path with regression coverage. To keep the PR mergeable against current `dev`, this also promotes `regex` to a direct `loongclaw-app` dependency because `turn_shared.rs` now uses it outside the `tool-websearch`-only path, and bumps workspace `tar` to `0.4.45` for the current RustSec advisories.
- What did not change (scope boundary): This PR does not introduce broader secret storage abstractions, provider-wide webhook rewrites, or additional issue `#424` / `#296` feature work.

## Linked Issues

- Closes #428
- Related #424

## Change Type

- [x] Bug fix
- [ ] Feature
- [ ] Refactor
- [ ] Documentation
- [x] Security hardening
- [ ] CI / workflow / release

## Touched Areas

- [ ] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [ ] Daemon / CLI / install
- [ ] Providers / routing
- [x] Tools
- [ ] Browser automation
- [x] Channels / integrations
- [ ] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [ ] Config / migration / onboarding
- [ ] Docs / contributor workflow
- [ ] CI / release / workflows

## Risk Track

- [x] Track A (routine / low-risk)
- [ ] Track B (higher-risk / policy-impacting)

If Track B, fill these in:

- Risk notes:
- Rollout / guardrails:
- Rollback path:

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `cargo test --workspace --locked`
- [x] `cargo test --workspace --all-features --locked`
- [x] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [ ] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [ ] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo fmt --all -- --check
- passed

cargo clippy --workspace --all-targets --all-features -- -D warnings
- passed

env -u ARK_API_KEY cargo test --workspace --locked
- passed

env -u ARK_API_KEY cargo test --workspace --all-features --locked
- passed

cargo audit
- passed after upgrading tar to 0.4.45

cargo check -p loongclaw-app --no-default-features --features channel-cli,channel-telegram,channel-feishu,channel-matrix,config-toml,memory-sqlite,feishu-integration,tool-shell,tool-file,tool-browser,provider-openai,provider-anthropic,provider-bedrock,provider-volcengine
- passed on the branch head

# merge-state verification in a temporary worktree after applying this patch on top of current origin/dev
cargo check -p loongclaw-app --no-default-features --features channel-cli,channel-telegram,channel-feishu,channel-matrix,config-toml,memory-sqlite,feishu-integration,tool-shell,tool-file,tool-browser,provider-openai,provider-anthropic,provider-bedrock,provider-volcengine
- passed on the simulated PR merge state

cargo test -p loongclaw-app --test crypto_helpers -- --nocapture
- passed

cargo test -p loongclaw-app feishu_token_mismatch_is_rejected -- --nocapture
- passed

cargo test -p loongclaw-app signature_verification_rejects_mismatch -- --nocapture
- passed

cargo test -p loongclaw-app tool_invoke_rejects_tampered_or_missing_leases -- --nocapture
- passed
```

## User-visible / Operator-visible Changes

- Feishu webhook token/signature checks and tool lease verification now use a shared timing-safe comparison path.
- `loongclaw-app` now exposes reusable HMAC-SHA256 verification helpers for follow-up webhook hardening work.
- No intended operator-visible behavior change beyond keeping the current `dev` merge path green and clearing the tar audit gate.

## Failure Recovery

- Fast rollback or disable path: revert commit `894c34b5` to drop the CI/audit compatibility follow-up, or revert `9c6b0596` and `894c34b5` together to restore the previous verification path entirely.
- Observable failure symptoms reviewers should watch for: legitimate Feishu callbacks or signed tool leases failing verification if an upstream caller depends on malformed or non-hex signature formatting, or unexpected build-matrix fallout if another feature still assumes `regex` is websearch-only.

## Reviewer Focus

- Confirm the shared helper semantics in `crates/app/src/crypto/` match the existing hex/string expectations at each call site.
- Check that the three migrated verification paths preserve current failure modes while removing plain equality on secrets/signatures.
- Confirm the direct `regex` dependency aligns with current `turn_shared.rs` usage on `dev`, and that the `tar` patch bump is an isolated security update rather than a behavioral change.

## Summary by CodeRabbit

* **Security**
  * Implemented constant-time byte comparisons across token and signature verification workflows to mitigate timing attack vulnerabilities.

* **New Features**
  * Added cryptographic utilities for HMAC-SHA256 computation and verification.

### CODE REVIEW

File: crates/app/Cargo.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ tool-shell = []
 tool-file = []
 tool-browser = ["dep:scraper"]
 tool-webfetch = []
-tool-websearch = ["dep:regex"]
+tool-websearch = []
```
[the remark this answers is not in the document]
Comment (chumyin) [reply crates/app/Cargo.toml:25]: Not applying this suggestion. `origin/dev` currently uses `regex::Regex` unconditionally in `crates/app/src/conversation/turn_shared.rs` for `strip_think_tags`, so making `regex` optional re-breaks the merge-state `rust-check-browser-without-webfetch` build. I re-verified that feature set against current `dev`, and this PR now includes an inline manifest comment documenting why `regex` stays unconditional until that call site is feature-gated.
